### PR TITLE
[BUILD] Introduce CXX 20 CI pipeline for MSVC/Windows

### PR DESCRIPTION
**Pull request by ThomsonTan**
Fixes # (issue)

## Changes

Please provide a brief description of the changes here.

For significant contributions please make sure you have completed the following items:

* [x] `CHANGELOG.md` updated for non-trivial changes
* [x] Unit tests have been added
* [ ] Changes in public API reviewed

### CODE REVIEW

**Review by lalitb — APPROVED** [review]

File: api/include/opentelemetry/std/span.h [hunk 1]
```diff
@@ -60,7 +60,7 @@ OPENTELEMETRY_END_NAMESPACE
 OPENTELEMETRY_BEGIN_NAMESPACE
 namespace nostd
 {
-constexpr std::size_t dynamic_extent = (std::numeric_limits<std::size_t>::max)();
+constexpr std::size_t dynamic_extent = std::numeric_limits<std::size_t>::max();
```
Comment (marcalff) [inline api/include/opentelemetry/std/span.h:63]: Why was this changed back, this was just fixed for NOMINMAX in windows ?

Comment (lalitb) [reply api/include/opentelemetry/std/span.h:63]: oh this was added after I approved the PR :)

Comment (ThomsonTan) [reply api/include/opentelemetry/std/span.h:63]: Apologizes for this. I was trying to see if the new CI could catch this kind of issue. Will revert this before merging.

Comment (ThomsonTan) [reply api/include/opentelemetry/std/span.h:63]: Converted this PR to draft for now.

Comment (ThomsonTan) [reply api/include/opentelemetry/std/span.h]: It is verified that the new CI pipeline can catch the issue. See the CI result below which reproduces the issue.

https://github.com/open-telemetry/opentelemetry-cpp/actions/runs/7204221666/job/19625317139?pr=2450

Comment (lalitb) [reply api/include/opentelemetry/std/span.h]: Thanks for validating this.

File: api/test/nostd/variant_test.cc [hunk 2]
```diff
@@ -48,7 +48,7 @@ TEST(VariantTest, Get)
   EXPECT_EQ(nostd::get<int>(w), 12);
   EXPECT_EQ(*nostd::get_if<int>(&v), 12);
   EXPECT_EQ(nostd::get_if<float>(&v), nullptr);
-#if __EXCEPTIONS
+#if __EXCEPTIONS || (defined(OPENTELEMETRY_STL_VERSION) && (OPENTELEMETRY_STL_VERSION >= 2020))
```
Comment (marcalff) [inline api/test/nostd/variant_test.cc:51]: std::variant is introduced in C++17, not C++20.

Please investigate (local CXX 17 MSVC build) if the logic should be `OPENTELEMETRY_STL_VERSION >= 2017` instead.

Also, somehow the existing code passed in C++14, C++17, C++20 builds in CI with GCC.

Not sure which platform and/or which compiler do set __EXCEPTIONS, please clarify if you know the context.

File: api/test/nostd/string_view_test.cc [hunk 3]
```diff
@@ -71,8 +71,8 @@ TEST(StringViewTest, SubstrPortion)
 TEST(StringViewTest, SubstrOutOfRange)
 {
   string_view s = "abc123";
-#if __EXCEPTIONS || ((defined(OPENTELEMETRY_STL_VERSION) && (OPENTELEMETRY_STL_VERSION >= 2020)))
-  EXPECT_THROW(s.substr(10), std::out_of_range);
+#if __EXCEPTIONS || (defined(OPENTELEMETRY_STL_VERSION) && (OPENTELEMETRY_STL_VERSION >= 2020))
+  EXPECT_THROW((void)s.substr(10), std::out_of_range);
```
Comment (marcalff) [inline api/test/nostd/string_view_test.cc:75]: I suspect the existing code was incorrect, as std::string_view is introduced in C++17, not C++20.

Please investigate (local CXX 17 MSVC build) if the logic should be `OPENTELEMETRY_STL_VERSION >= 2017` instead.

**Review by marcalff — COMMENTED** [review]
Thanks for the better CI coverage.

See comments on unit tests, about C++17 vs C++20.

File: api/test/nostd/variant_test.cc (hunk 2 above)
Comment (marcalff) [reply api/test/nostd/variant_test.cc:51]: Found out that __EXCEPTIONS is set by GCC and Clang, except when building with -fno-exceptions.

This is likely why a GCC build with exceptions passed this test.

Comment (ThomsonTan) [reply api/test/nostd/variant_test.cc:51]: Thanks for looking into it. Will try MSVC C++ 17 build.

File: api/test/nostd/string_view_test.cc (hunk 3 above)
Comment (ThomsonTan) [reply api/test/nostd/string_view_test.cc]: Verified that std::out_of_range exception was thrown out with MSVC/C++17. Thanks for the catch.

**Review by marcalff — APPROVED** [review]
